gtsam/linear/KalmanFilter.cpp

```diff
@@ -63,20 +71,27 @@ KalmanFilter::fuse(const State& p, GaussianFactor::shared_ptr newFactor) const {
 
 /* ************************************************************************* */
 KalmanFilter::State KalmanFilter::init(const Vector& x0,
-    const SharedDiagonal& P0) const {
-
+                                       const SharedDiagonal& P0) const {
   // Create a factor graph f(x0), eliminate it into P(x0)
   GaussianFactorGraph factorGraph;
-  factorGraph.emplace_shared<JacobianFactor>(0, I_, x0, P0); // |x-x0|^2_diagSigma
+  factorGraph.emplace_shared<JacobianFactor>(0, I_, x0,
+                                             P0);  // |x-x0|^2_P0
   return solve(factorGraph);
 }
 
 /* ************************************************************************* */
-KalmanFilter::State KalmanFilter::init(const Vector& x, const Matrix& P0) const {
-
+KalmanFilter::State KalmanFilter::init(const Vector& x0,
+                                       const Matrix& P0) const {
   // Create a factor graph f(x0), eliminate it into P(x0)
   GaussianFactorGraph factorGraph;
-  factorGraph.emplace_shared<HessianFactor>(0, x, P0); // 0.5*(x-x0)'*inv(Sigma)*(x-x0)
+
+  // Perform Cholesky decomposition of P0
+  FACTORIZE_Q_INTO_L(P0)
```

Comment:
In all honesty, I am not a big fan of this. It's not functional and the variable `L` seems to magically appear, making the code less readable.

I did some investigating and perhaps we can instead define the following function
```cpp
Matrix factorizeQIntoR(const Matrix& Q) {
  Eigen::LLT<Matrix> llt(Q);
  auto L = llt.matrixL();
  Matrix I = Matrix::Identity(Q.rows(), Q.cols());
  return L.solve(I);
}
```

and you can use it as
```cpp
const Matrix R = factorizeQIntoR(P0);
const Vector b = R * x0;  // = R*x0
```
which makes it self-documenting as a bonus.

Reply:
One more thing: you can do `R.triangularView<Eigen::Lower>() * x0` to improve performance since it takes advantage of the fact that R is lower triangular.

I can't seem to find a way to return an `Eigen::TriangularView<const Matrix, Eigen::Lower>` in a way that works,  which I guess is why you went the macro route in the first place.

Reply:
R is upper triangular. L.solve(x) is same as multiplying with R, read comment more closely :-)

Reply:
I can’t use your function as I need to solve multiple times with same L. Storing R is less efficient.
Is the macro a dealbreaker? It works, it’s faster, it avoids hard Eigen type issues.

Reply:
Then I found something interesting. I printed `L.solve(I)` to verify things when coming up with the above, and `R` is lower triangular.

Regarding the macro, I am not opposed to it, I was hoping for a cleaner, less magicky solution. :)

Reply:
Oh, my, this was a rabbithole, but indeed. $R^TR = L^{-T}L^{-1}$ does *not* imply $R = L^{-1}$. I updated the comment and found an elegant way to not have a macro:

```c++
// In the code below we often get a covariance matrix Q, and we need to create a
// JacobianFactor with cost 0.5 * |Ax - b|^T Q^{-1} |Ax - b|. Factorizing Q as
//   Q = L L^T
// and hence
//   Q^{-1} = L^{-T} L^{-1} = M^T M, with M = L^{-1}
// We then have
//   0.5 * |Ax - b|^T Q^{-1} |Ax - b|
// = 0.5 * |Ax - b|^T M^T M |Ax - b|
// = 0.5 * |MAx - Mb|^T |MAx - Mb|
// The functor below efficiently multiplies with M by calling L.solve().

struct InverseL : public Eigen::LLT<Matrix> {
  InverseL(const Matrix& Q) : Eigen::LLT<Matrix>(Q) {}
  Matrix operator*(const Matrix& A) const { return matrixL().solve(A); }
};
```